packages/gatsby-source-contentful/src/extend-node-type.js

```diff
@@ -542,18 +541,83 @@ const fluidNodeType = ({ name, getTracedSVG }) => {
   }
 }
 
-exports.extendNodeType = ({ type, store }) => {
+exports.extendNodeType = ({ type, store, cache }) => {
   if (type.name.match(/contentful.*RichTextNode/)) {
     return {
       nodeType: {
         type: GraphQLString,
-        deprecationReason: `This field is deprecated, please use 'json' instead.`,
+        deprecationReason: `This field is deprecated, please use 'raw' instead. @todo add link to migration steps.`,
       },
       json: {
-        type: GraphQLJSON,
-        resolve: (source, fieldArgs) => {
-          const contentJSON = JSON.parse(source.internal.content)
-          return contentJSON
+        type: GraphQLString,
+        deprecationReason: `This field is deprecated, please use 'raw' instead. @todo add link to migration steps.`,
+      },
+      references: {
+        type: [`ContentfulReference`],
+        async resolve(source, args, context, info) {
+          const parent = await context.nodeModel.findRootNodeAncestor(source)
+
+          const rawReferences = { Entry: [], Asset: [] }
+
+          // Locate all Contentful Links within the rich text data
+          const traverse = obj => {
+            for (let k in obj) {
+              const value = obj[k]
+              if (value && value.sys && value.sys.type === `Link`) {
+                rawReferences[value.sys.linkType].push(value.sys.contentful_id)
+              } else if (value && typeof value === `object`) {
+                traverse(value)
+              }
+            }
+          }
+
+          traverse(JSON.parse(source.raw))
+
+          if (!rawReferences.Entry.length && !rawReferences.Asset.length) {
+            return null
```

Comment:
Maybe return an empty array instead? Or do you have a reason to explicitly return `null` here?

Reply:
Sure, makes sense to return an empty list if nothing is found.

Just need to adjust the check here as well:
https://github.com/gatsbyjs/gatsby/pull/24905/files#diff-600985b26047b64df633c2b0157343e0R8